velox/type/DecimalUtil.h

```diff
@@ -193,6 +194,44 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  /// Rescale a double value to decimal value.
+  ///
+  /// Use `folly::tryTo` to convert a double value to int128_t or int64_t. It
+  /// returns an error when overflow occurs so that we could determine whether
+  /// an overflow occurs through checking the result.
+  ///
+  /// Normally, return the rescaled value. Otherwise, if the `toValue` overflows
+  /// the TOutput's limits or the `toValue` exceeds the precision's limits, it
+  /// will set an error message in `error` field and return an std::nullopt.
+  template <typename TOutput>
+  inline static std::optional<TOutput>
+  rescaleDouble(double value, int precision, int scale, std::string& error) {
```

Comment:
I'm seeing multiple PRs starting to introduce similar APIs. I wonder if we should introduce something similar to StatusOr or whatever is used by folly::tryTo?

Reply:
I'm not familiar with Folly's APIs. But for `folly::tryTo`, it will return a value whose type is `folly::Expected`. Maybe we could use `folly::Expected` as a return value instead of an optional return value and an optional error.
Let me have a try.

Reply:
@mbasmanova I just tried to use `folly::Expected` as the return value's type. Do you think that makes sense?

Reply:
Something like this might work. I wonder about the usage of std::string for error though. Looks like this may result in a lots of copying of strings. I wonder if we should use an enum instead.

Reply:
done

Reply:
It won't copy strings, we use std::string& reference, and pass it to makeBadCastException